README
NAME
    AnyEvent::HTTPD - A simple lightweight event based web (application)
    server

VERSION
    Version 0.82

SYNOPSIS
        use AnyEvent::HTTPD;

        my $httpd = AnyEvent::HTTPD->new (port => 9090);

        $httpd->reg_cb (
           '/' => sub {
              my ($httpd, $req) = @_;

              $req->respond ({ content => ['text/html',
                 "<html><body><h1>Hello World!</h1>"
                 . "<a href=\"/test\">another test page</a>"
                 . "</body></html>"
              ]});
           },
           '/test' => sub {
              my ($httpd, $req) = @_;

              $req->respond ({ content => ['text/html',
                 "<html><body><h1>Test page</h1>"
                 . "<a href=\"/\">Back to the main page</a>"
                 . "</body></html>"
              ]});
           },
        );

        $httpd->run; # making a AnyEvent condition variable would also work

DESCRIPTION
    This module provides a simple HTTPD for serving simple web application
    interfaces. It's completly event based and independend from any event
    loop by using the AnyEvent module.

    It's HTTP implementation is a bit hacky, so before using this module
    make sure it works for you and the expected deployment. Feel free to
    improve the HTTP support and send in patches!

    The documentation is currently only the source code, but next versions
    of this module will be better documented hopefully. See also the
    "samples/" directory in the AnyEvent::HTTPD distribution for basic
    starting points.

FEATURES
    *   support for GET and POST requests.

    *   support for HTTP 1.0 keep-alive.

    *   processing of "x-www-form-urlencoded" and "multipart/form-data"
        ("multipart/mixed") encoded form parameters.

    *   support for streaming responses.

    *   with version 0.8 no more dependend on LWP for HTTP::Date.

METHODS
    The AnyEvent::HTTPD class inherits directly from
    AnyEvent::HTTPD::HTTPServer which inherits the event callback interface
    from Object::Event.

    Event callbacks can be registered via the Object::Event API (see the
    documentation of Object::Event for details).

    For a list of available events see below in the *EVENTS* section.

    new (%args)
        This is the constructor for a AnyEvent::HTTPD object. The %args hash
        may contain one of these key/value pairs:

        host => $host
            The TCP address of the HTTP server will listen on. Usually
            0.0.0.0 (the default), for a public server, or 127.0.0.1 for a
            local server.

        port => $port
            The TCP port the HTTP server will listen on. If undefined some
            free port will be used. You can get it via the "port" method.

        request_timeout => $seconds
            This will set the request timeout for connections. The default
            value is 60 seconds.

    port
        Returns the port number this server is bound to.

    host
        Returns the host/ip this server is bound to.

    stop_request
        When the server walks the request URI path upwards you can stop the
        walk by calling this method. You can even stop further handling
        after the "request" event.

        Example:

           $httpd->reg_cb (
              '/test' => sub {
                 my ($httpd, $req) = @_;

                 # ...

                 $httpd->stop_request; # will prevent that the callback below is called
              },
              '' => sub { # this one wont be called by a request to '/test'
                 my ($httpd, $req) = @_;

                 # ...
              }
           );

    run This method is a simplification of the "AnyEvent" condition variable
        idiom. You can use it instead of writing:

           my $cvar = AnyEvent->condvar;
           $cvar->wait;

    stop
        This will stop the HTTP server and return from the "run" method if
        you started the server via that method!

EVENTS
    Every request goes to a specific URL. After a (GET or POST) request is
    received the URL's path segments are walked down and for each segment a
    event is generated. An example:

    If the URL '/test/bla.jpg' is requestes following events will be
    generated:

      '/test/bla.jpg' - the event for the last segment
      '/test'         - the event for the 'test' segment
      ''              - the root event of each request

    To actually handle any request you just have to register a callback for
    the event name with the empty string. To handle all requests in the
    '/test' directory you have to register a callback for the event with the
    name '/test'. Here is an example how to register an event for the
    example URL above:

       $httpd->reg_cb (
          '/test/bla.jpg' => sub {
             my ($httpd, $req) = @_;

             $req->respond ([200, 'ok', { 'Content-Type' => 'text/html' }, '<h1>Test</h1>' }]);
          }
       );

    See also "stop_request" about stopping the walk of the path segments.

    The first argument to such a callback is always the AnyEvent::HTTPD
    object itself. The second argument ($req) is the
    AnyEvent::HTTPD::Request object for this request. It can be used to get
    the (possible) form parameters for this request or the transmitted
    content and respond to the request.

    Also every request also emits the "request" event, with the same
    arguments and semantics, you can use this to implement your own request
    multiplexing.

CACHING
    Any response from the HTTP server will have "Cache-Control" set to
    "max-age=0" and also the "Expires" header set to the "Date" header.
    Meaning: Caching is disabled.

    If you need caching or would like to have it you can send me a mail or
    even better: a patch :)

AUTHOR
    Robin Redeker, "<elmex at ta-sa.org>"

BUGS
    Please report any bugs or feature requests to "bug-bs-httpd at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=AnyEvent-HTTPD>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc AnyEvent::HTTPD

    You can also look for information at:

    *   Git repository

        <http://git.ta-sa.org/AnyEvent-HTTPD.git>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=AnyEvent-HTTPD>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/AnyEvent-HTTPD>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/AnyEvent-HTTPD>

    *   Search CPAN

        <http://search.cpan.org/dist/AnyEvent-HTTPD>

ACKNOWLEDGEMENTS
       Andrey Smirnov - for keep-alive patches.
       Pedro Melo     - for valuable input in general and patches.

COPYRIGHT & LICENSE
    Copyright 2008-2009 Robin Redeker, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for AnyEvent-HTTPD

{{$NEXT}}

	- fixed use statements in the tests.

0.93    Thu Aug  4 10:38:09 CEST 2011
        - made it possible to remove default headers
          completely, thanks go to Chris Kastorff for providing a patch.
        - applied also a fix for case handling of the headers,
          thanks go to the patch from Chris Kastorff.
        - fixed some tests. I hate it when all bugs that
          are found via CPAN Testers are bugs in the tests.
        - fixed that the URL returned by Request::url method will
          now also contain the query string (thanks to art sackett for
          spotting this).

0.92    Tue Mar  1 11:46:14 CET 2011
        - added simple SSL/TLS support.
        - removed from tests from the distribution.
        - replaced 01_basic_request.t with a non-wget variant for CPAN.

0.91    Wed Feb  2 15:21:02 CET 2011
        - set the Date header correctly.
        - append a newline to error messages.
        - hopefully improved connection closing mechanism.
        - don't send body when responding to a HEAD request.

0.90    Mon Jan 31 13:04:45 CET 2011
        - fixed a small thing with response on dead connections.
        - added small example of 'request' event usage.
        - make Content-Length: not be empty.
        - hand HEAD requests down to the event handlers.
        - added allowed_methods support from nrh.
        - don't force cache policy anymore.

0.84    Mon Nov  9 17:18:07 CET 2009
        - added client_connected and client_disconnected events.
        - returning 501 instead of 405 on unimplemented method.
        - optimized regexes (contributed by Mons Anderson).
        - added backlog option to AE::HTTPD (contributed by Mons Anderson).

0.83    Tue Sep 29 16:29:57 CEST 2009
        - added client host and port to the request.

0.82    Tue Aug 11 08:33:05 CEST 2009
        - added common::sense to prereq.
        - made tests using 127.0.0.1.

0.81    Mon Aug 10 09:59:16 CEST 2009
        - cleaned up dependencies in Makefile.PL.
        - using common::sense in all modules & scripts.

0.8     Sun Aug  9 18:10:23 CEST 2009
        - made tests run with AnyEvent::Impl::Perl.
        - removed HTTP::Date (LWP) dependency.

0.71    Sat Aug  8 00:43:58 CEST 2009
        - fixed the published tests. they used Coro.

0.7     Fri Aug  7 15:37:27 CEST 2009
        - fixed documentation in AE::HTTP::Request.
        - added asynchronous response functionality,
          for sending large files via AnyEvent::AIO for instance.
        - added tests to distribution (instead of maintainer only tests).
        - fixed a bug with reception of requests without headers.

0.6     Tue Jul 14 11:16:44 CEST 2009
        - fixed leaking AE::HTTPD::HTTPConnection.

0.5     Thu Jul  2 04:30:14 CEST 2009
        - applied patch from mathieu at closetwork.org to add a
          host parameter for binding.
        - removed bogus API stuff like ::Appgets or the weird
          form and response handling. AnyEvent::HTTPD should be and
          stay a simple HTTP server for simple purposes.
          If you need further sugar, please write your own modules for
          it. If you need anything ask me or look in the git repository
          at http://git.ta-sa.org/AnyEvent-HTTPD.git
        - added responded method to ::Request.
        - stop_request now also stops further handling of the request
          after the 'request' event.
        - added 'request_timeout's
        - added keep-alive support (for HTTP 1.0) (thanks to Andrey Smirnov).

0.04    Sun Dec 28 15:48:28 CET 2008
        - removed TCP* classes and using AnyEvent::Handle instead.
        - added size and maxlength args to the C<entry> function in Appgets
        - changed the API to actually call the events for all path segments
          of an URL. also removed the ugly '/' => '_' mapping for the path
          seperators. Sorry for any breakage in your code ;-/
        - removed Perl 5.10 dependency.

0.03    Tue Apr 15 12:57:10 CEST 2008
        - added submit() function to Appgets.
        - fixed a bug in the url() method of AnyEvent::HTTPD::Request where
          the url wasn't get correctly. resulting in bad form behaviour.

0.02    Sat Mar 22 16:09:55 CET 2008
        - rewrote the module API to be more "async".
        - made it possible to respond to requests asynchronously
          in C<AnyEvent::HTTPD>.
        - added lots of example scripts.

0.01    Thu Mar 20 19:53:31 CET 2008
        - first release.


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'AnyEvent::HTTPD',
    AUTHOR              => 'Robin Redeker <elmex@ta-sa.org>',
    VERSION_FROM        => 'lib/AnyEvent/HTTPD.pm',
    ABSTRACT_FROM       => 'lib/AnyEvent/HTTPD.pm',
    LICENSE             => 'perl',
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More'     => 0,
        'AnyEvent'       => 0,
        'Object::Event'  => 0,
        'URI'            => 0,
        'Time::Local'    => 0,
        'common::sense'  => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz',
                             PREOP => 'pod2text lib/AnyEvent/HTTPD.pm | tee README >$(DISTVNAME)/README; chmod -R u=rwX,go=rX . ;',
    },
    clean               => { FILES => 'AnyEvent-HTTPD-*' },
);


lib/AnyEvent/HTTPD/Util.pm
package AnyEvent::HTTPD::Util;
use AnyEvent;
use AnyEvent::Socket;
use common::sense;

require Exporter;
our @ISA = qw/Exporter/;

our @EXPORT = qw/parse_urlencoded url_unescape header_set
                 header_get header_exists/;

=head1 NAME

AnyEvent::HTTPD::Util - Utility functions for AnyEvent::HTTPD

=head1 SYNOPSIS

=head1 DESCRIPTION

The functions in this package are not public.

=over 4

=cut

sub url_unescape {
   my ($val) = @_;
   $val =~ s/\+/\040/g;
   $val =~ s/%([0-9a-fA-F][0-9a-fA-F])/chr (hex ($1))/eg;
   $val
}

sub parse_urlencoded {
   my ($cont) = @_;
   my (@pars) = split /[\&\;]/, $cont;
   $cont = {};

   for (@pars) {
      my ($name, $val) = split /=/, $_;
      $name = url_unescape ($name);
      $val  = url_unescape ($val);

      push @{$cont->{$name}}, [$val, ''];
   }
   $cont
}

sub test_connect {
   my ($host, $port, $data) = @_;

   my $c = AE::cv;

   my $t; $t = AnyEvent->timer (after => 0.1, cb => sub {
      my $hdl;
      my $buf;
      undef $t;
      tcp_connect $host, $port, sub {
         my ($fh) = @_
            or die "couldn't connect: $!";

         $hdl =
            AnyEvent::Handle->new (
               fh => $fh,
               timeout => 15,
               on_eof => sub {
                  $c->send ($buf);
                  undef $hdl;
               },
               on_timeout => sub {
                  warn "test_connect timed out";
                  $c->send ($buf);
                  undef $hdl;
               },
               on_read => sub {
                  $buf .= $hdl->rbuf;
                  $hdl->rbuf = '';
               });
         $hdl->push_write ($data);
      };
   });

   $c
}

###
# these functions set/get/check existence of a header name:value pair while
# ignoring the case of the name
#
# quick hack, does not scale to large hashes. however, it's not expected to be
# run on large hashes.
#
# a more performant alternative would be to keep two hashes for each set of
# headers, one for the headers in the case they like, and one a mapping of
# names from some consistent form (say, all lowercase) to the name in the other
# hash, including capitalization. (this style is used in HTTP::Headers)

sub _header_transform_case_insens {
   my $lname = lc $_[1];
   my (@names) = grep { $lname eq lc ($_) } keys %{$_[0]};
   @names ? $names[0] : $_[1]
}

sub header_set {
    my ($hdrs, $name, $value) = @_;
    $name = _header_transform_case_insens ($hdrs, $name);
    $hdrs->{$name} = $value;
}

sub header_get {
    my ($hdrs, $name) = @_;
    $name = _header_transform_case_insens ($hdrs, $name);
    exists $hdrs->{$name} ? $hdrs->{$name} : undef
}

sub header_exists {
    my ($hdrs, $name) = @_;
    $name = _header_transform_case_insens ($hdrs, $name);
    # NB: even if the value is undefined, return true
    return exists $hdrs->{$name}
}

=back

=head1 AUTHOR

Robin Redeker, C<< <elmex@ta-sa.org> >>

=head1 SEE ALSO

=head1 COPYRIGHT & LICENSE

Copyright 2009 Robin Redeker, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;



lib/AnyEvent/HTTPD/Request.pm
package AnyEvent::HTTPD::Request;
use common::sense;

=head1 NAME

AnyEvent::HTTPD::Request - A web application request handle for L<AnyEvent::HTTPD>

=head1 DESCRIPTION

This is the request object as generated by L<AnyEvent::HTTPD> and given
in the request callbacks.

=head1 METHODS

=over 4

=cut

sub new {
   my $this  = shift;
   my $class = ref($this) || $this;
   my $self  = { @_ };
   bless $self, $class
}

=item B<url>

This method returns the URL of the current request as L<URI> object.

=cut

sub url {
   my ($self) = @_;
   my $url = $self->{url};
   my $u = URI->new ($url);
   $u
}

=item B<respond ([$res])>

C<$res> can be:

=over 4

=item * an array reference

Then the array reference has these elements:

   my ($code, $message, $header_hash, $content) =
         [200, 'ok', { 'Content-Type' => 'text/html' }, '<h1>Test</h1>' }]

You can remove most headers added by default (like C<Cache-Control>,
C<Expires>, and C<Content-Length>) by setting them to undef, like so:

   $req->respond([
      200, 'OK', {
        'Content-Type'  => 'text/html',
        'Cache-Control' => 'max-age=3600',
        'Expires'       => undef,
      },
      'This data will be cached for one hour.'
   ]);

=item * a hash reference

If it was a hash reference the hash is first searched for the C<redirect>
key and if that key does not exist for the C<content> key.

The value for the C<redirect> key should contain the URL that you want to redirect
the request to.

The value for the C<content> key should contain an array reference with the first
value being the content type and the second the content.

=back

Here is an example:

   $httpd->reg_cb (
      '/image/elmex' => sub {
         my ($httpd, $req) = @_;

         open IMG, "$ENV{HOME}/media/images/elmex.png"
            or $req->respond (
                  [404, 'not found', { 'Content-Type' => 'text/plain' }, 'not found']
               );

         $req->respond ({ content => ['image/png', do { local $/; <IMG> }] });
      }
   );

B<How to send large files:>

For longer responses you can give a callback instead of a string to
the response function for the value of the C<$content>.

   $req->respond ({ content => ['video/x-ms-asf', sub {
      my ($data_cb) = @_;

      # start some async retrieve operation, for example use
      # IO::AIO (with AnyEvent::AIO). Or retrieve chunks of data
      # to send somehow else.

   } });

The given callback will receive as first argument either another callback
(C<$data_cb> in the above example) or an undefined value, which means that
there is no more data required and the transfer has been completed (either by
you sending no more data, or by a disconnect of the client).

The callback given to C<respond> will be called whenever the send queue of the
HTTP connection becomes empty (meaning that the data is written out to the
kernel). If it is called you have to start delivering the next chunk of data.

That doesn't have to be immediately, before the callback returns.  This means
that you can initiate for instance an L<IO::AIO> request (see also
L<AnyEvent::AIO>) and send the data later.  That is what the C<$data_cb>
callback is for. You have to call it once you got the next chunk of data. Once
you sent a chunk of data via C<$data_cb> you can just wait until your callback
is called again to deliver the next chunk.

If you are done transferring all data call the C<$data_cb> with an empty string
or with no argument at all.

Please consult the example script C<large_response_example> from the
C<samples/> directory of the L<AnyEvent::HTTPD> distribution for an example of
how to use this mechanism.

B<NOTE:> You should supply a 'Content-Length' header if you are going to send a
larger file. If you don't do that the client will have no chance to know if the
transfer was complete. To supply additional header fields the hash argument
format will not work. You should use the array argument format for this case.

=cut

sub respond {
   my ($self, $res) = @_;

   return unless $self->{resp};

   my $rescb = delete $self->{resp};

   if (ref $res eq 'HASH') {
      my $h = $res;
      if ($h->{redirect}) {
         $res = [
            301, 'redirected', { Location => $h->{redirect} },
            "Redirected to <a href=\"$h->{redirect}\">here</a>"
         ];
      } elsif ($h->{content}) {
         $res = [
            200, 'ok', { 'Content-Type' => $h->{content}->[0] },
            $h->{content}->[1]
         ];
      }

   }

   $self->{responded} = 1;

   my $no_body = $self->method eq 'HEAD';

   if (not defined $res) {
      $rescb->(404, "ok", { 'Content-Type' => 'text/html' }, "<h1>No content</h1>", $no_body);

   } else {
      $rescb->(@$res, $no_body);
   }
}

=item B<responded>

Returns true if this request already has been responded to.

=cut

sub responded { $_[0]->{responded} }

=item B<parm ($key)>

Returns the first value of the form parameter C<$key> or undef.

=cut

sub parm {
   my ($self, $key) = @_;

   if (exists $self->{parm}->{$key}) {
      return $self->{parm}->{$key}->[0]->[0]
   }

   return undef;
}

=item B<params>

Returns list of parameter names.

=cut

sub params { keys %{$_[0]->{parm} || {}} }

=item B<vars>

Returns a hash of form parameters. The value is either the 
value of the parameter, and in case there are multiple values
present it will contain an array reference of values.

=cut

sub vars {
   my ($self) = @_;

   my $p = $self->{parm};

   my %v = map {
      my $k = $_;
      $k =>
         @{$p->{$k}} > 1
            ? [ map { $_->[0] } @{$p->{$k}} ]
            : $p->{$k}->[0]->[0]
   } keys %$p;

   %v
}

=item B<method>

This method returns the method of the current request.

=cut

sub method { $_[0]{method} }

=item B<content>

Returns the request content or undef if only parameters for a form
were transmitted.

=cut

sub content { $_[0]->{content} }

=item B<headers>

This method will return a hash reference containing the HTTP headers for this
HTTP request.

=cut

sub headers { $_[0]->{hdr} }

=item B<client_host>

This method returns the host/IP of the HTTP client this request was received
from.

=cut

sub client_host { $_[0]->{host} }

=item B<client_port>

This method returns the TCP port number of the HTTP client this
request was received from.

=cut

sub client_port { $_[0]->{port} }

=back

=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Robin Redeker, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1;


lib/AnyEvent/HTTPD.pm
package AnyEvent::HTTPD;
use common::sense;
use Scalar::Util qw/weaken/;
use URI;
use AnyEvent::HTTPD::Request;
use AnyEvent::HTTPD::Util;

use base qw/AnyEvent::HTTPD::HTTPServer/;

=head1 NAME

AnyEvent::HTTPD - A simple lightweight event based web (application) server

=head1 VERSION

Version 0.93

=cut

our $VERSION = '0.93';

=head1 SYNOPSIS

    use AnyEvent::HTTPD;

    my $httpd = AnyEvent::HTTPD->new (port => 9090);

    $httpd->reg_cb (
       '/' => sub {
          my ($httpd, $req) = @_;

          $req->respond ({ content => ['text/html',
             "<html><body><h1>Hello World!</h1>"
             . "<a href=\"/test\">another test page</a>"
             . "</body></html>"
          ]});
       },
       '/test' => sub {
          my ($httpd, $req) = @_;

          $req->respond ({ content => ['text/html',
             "<html><body><h1>Test page</h1>"
             . "<a href=\"/\">Back to the main page</a>"
             . "</body></html>"
          ]});
       },
    );

    $httpd->run; # making a AnyEvent condition variable would also work

=head1 DESCRIPTION

This module provides a simple HTTPD for serving simple web application
interfaces. It's completly event based and independend from any event loop
by using the L<AnyEvent> module.

It's HTTP implementation is a bit hacky, so before using this module make sure
it works for you and the expected deployment. Feel free to improve the HTTP
support and send in patches!

The documentation is currently only the source code, but next versions of this
module will be better documented hopefully. See also the C<samples/> directory
in the L<AnyEvent::HTTPD> distribution for basic starting points.

=head1 FEATURES

=over 4

=item * support for GET and POST requests.

=item * support for HTTP 1.0 keep-alive.

=item * processing of C<x-www-form-urlencoded> and C<multipart/form-data> (C<multipart/mixed>) encoded form parameters.

=item * support for streaming responses.

=item * with version 0.8 no more dependend on L<LWP> for L<HTTP::Date>.

=item * (limited) support for SSL

=back

=head1 METHODS

The L<AnyEvent::HTTPD> class inherits directly from
L<AnyEvent::HTTPD::HTTPServer> which inherits the event callback interface from
L<Object::Event>.

Event callbacks can be registered via the L<Object::Event> API (see the
documentation of L<Object::Event> for details).

For a list of available events see below in the I<EVENTS> section.

=over 4

=item B<new (%args)>

This is the constructor for a L<AnyEvent::HTTPD> object.
The C<%args> hash may contain one of these key/value pairs:

=over 4

=item host => $host

The TCP address of the HTTP server will listen on. Usually 0.0.0.0 (the
default), for a public server, or 127.0.0.1 for a local server.

=item port => $port

The TCP port the HTTP server will listen on. If undefined some
free port will be used. You can get it via the C<port> method.

=item ssl => $tls_ctx

If this option is given the server will listen for a SSL/TLS connection on the
configured port. As C<$tls_ctx> you can pass anything that you can pass as
C<tls_ctx> to an L<AnyEvent::Handle> object.

Example:

   my $httpd =
      AnyEvent::HTTPD->new (
         port => 443,
         ssl  => { cert_file => "/path/to/my/server_cert_and_key.pem" }
      );

Or:

   my $httpd =
      AnyEvent::HTTPD->new (
         port => 443,
         ssl  => AnyEvent::TLS->new (...),
      );

=item request_timeout => $seconds

This will set the request timeout for connections.
The default value is 60 seconds.

=item backlog => $int

The backlog argument defines the maximum length the queue of pending
connections may grow to.  The real maximum queue length will be 1.5 times more
than the value specified in the backlog argument.

See also C<man 2 listen>.

By default will be set by L<AnyEvent::Socket>C<::tcp_server> to C<128>.

=item connection_class => $class

This is a special parameter that you can use to pass your own connection class
to L<AnyEvent::HTTPD::HTTPServer>.  This is only of interest to you if you plan
to subclass L<AnyEvent::HTTPD::HTTPConnection>.

=item request_class => $class

This is a special parameter that you can use to pass your own request class
to L<AnyEvent::HTTPD>.  This is only of interest to you if you plan
to subclass L<AnyEvent::HTTPD::Request>.

=item allowed_methods => $arrayref

This parameter sets the allowed HTTP methods for requests, defaulting to GET,
HEAD and POST.  Each request received is matched against this list, and a
'501 not implemented' is returned if no match is found.  Requests using
disallowed handlers will never trigger callbacks.

=back

=cut

sub new {
   my $this  = shift;
   my $class = ref($this) || $this;
   my $self  = $class->SUPER::new (
      request_class => "AnyEvent::HTTPD::Request",
      @_
   );

   $self->reg_cb (
      connect => sub {
         my ($self, $con) = @_;

         weaken $self;

         $self->{conns}->{$con} = $con->reg_cb (
            request => sub {
               my ($con, $meth, $url, $hdr, $cont) = @_;
               #d# warn "REQUEST: $meth, $url, [$cont] " . join (',', %$hdr) . "\n";

               $url = URI->new ($url);

               if ($meth eq 'GET') {
                  $cont = parse_urlencoded ($url->query);
               }

               if ( scalar grep { $meth eq $_ } @{ $self->{allowed_methods} } ) {

                  weaken $con;

                  $self->handle_app_req (
                     $meth, $url, $hdr, $cont, $con->{host}, $con->{port},
                     sub {
                        $con->response (@_) if $con;
                     });
               } else {
                  $con->response (200, "ok");
               }
            }
         );

         $self->event (client_connected => $con->{host}, $con->{port});
      },
      disconnect => sub {
         my ($self, $con) = @_;
         $con->unreg_cb (delete $self->{conns}->{$con});
         $self->event (client_disconnected => $con->{host}, $con->{port});
      },
   );

   $self->{state} ||= {};

   return $self
}

sub handle_app_req {
   my ($self, $meth, $url, $hdr, $cont, $host, $port, $respcb) = @_;

   my $req =
      $self->{request_class}->new (
         httpd   => $self,
         method  => $meth,
         url     => $url,
         hdr     => $hdr,
         parm    => (ref $cont ? $cont : {}),
         content => (ref $cont ? undef : $cont),
         resp    => $respcb,
         host    => $host,
         port    => $port,
      );

   $self->{req_stop} = 0;
   $self->event (request => $req);
   return if $self->{req_stop};

   my @evs;
   my $cururl = '';
   for my $seg ($url->path_segments) {
      $cururl .= $seg;
      push @evs, $cururl;
      $cururl .= '/';
   }

   for my $ev (reverse @evs) {
      $self->event ($ev => $req);
      last if $self->{req_stop};
   }
}

=item B<port>

Returns the port number this server is bound to.

=item B<host>

Returns the host/ip this server is bound to.

=item B<allowed_methods>

Returns an arrayref of allowed HTTP methods, possibly as set by the
allowed_methods argument to the constructor.

=item B<stop_request>

When the server walks the request URI path upwards you can stop
the walk by calling this method. You can even stop further handling
after the C<request> event.

Example:

   $httpd->reg_cb (
      '/test' => sub {
         my ($httpd, $req) = @_;

         # ...

         $httpd->stop_request; # will prevent that the callback below is called
      },
      '' => sub { # this one wont be called by a request to '/test'
         my ($httpd, $req) = @_;

         # ...
      }
   );

=cut

sub stop_request {
   my ($self) = @_;
   $self->{req_stop} = 1;
}

=item B<run>

This method is a simplification of the C<AnyEvent> condition variable
idiom. You can use it instead of writing:

   my $cvar = AnyEvent->condvar;
   $cvar->wait;

=cut

sub run {
   my ($self) = @_;
   $self->{condvar} = AnyEvent->condvar;
   $self->{condvar}->wait;
}

=item B<stop>

This will stop the HTTP server and return from the
C<run> method B<if you started the server via that method!>

=cut

sub stop { $_[0]->{condvar}->broadcast if $_[0]->{condvar} }

=back

=head1 EVENTS

Every request goes to a specific URL. After a (GET or POST) request is
received the URL's path segments are walked down and for each segment
a event is generated. An example:

If the URL '/test/bla.jpg' is requestes following events will be generated:

  '/test/bla.jpg' - the event for the last segment
  '/test'         - the event for the 'test' segment
  ''              - the root event of each request

To actually handle any request you just have to register a callback for the event
name with the empty string. To handle all requests in the '/test' directory
you have to register a callback for the event with the name C<'/test'>.
Here is an example how to register an event for the example URL above:

   $httpd->reg_cb (
      '/test/bla.jpg' => sub {
         my ($httpd, $req) = @_;

         $req->respond ([200, 'ok', { 'Content-Type' => 'text/html' }, '<h1>Test</h1>' }]);
      }
   );

See also C<stop_request> about stopping the walk of the path segments.

The first argument to such a callback is always the L<AnyEvent::HTTPD> object
itself.  The second argument (C<$req>) is the L<AnyEvent::HTTPD::Request>
object for this request. It can be used to get the (possible) form parameters
for this request or the transmitted content and respond to the request.


Along with the above mentioned events these events are also provided:

=over 4

=item request => $req

Every request also emits the C<request> event, with the same arguments and
semantics as the above mentioned path request events.  You can use this to
implement your own request multiplexing. You can use C<stop_request> to stop
any further processing of the request as the C<request> event is the first
thing that is executed for an incoming request.

An example of one of many possible uses:

   $httpd->reg_cb (
      request => sub {
         my ($httpd, $req) = @_;

         my $url = $req->url;

         if ($url->path =~ /\/images\/img_(\d+).jpg$/) {
            handle_image_request ($req, $1); # your task :)

            # stop the request from emitting further events
            # so that the '/images/img_001.jpg' and the
            # '/images' and '' events are NOT emitted:
            $httpd->stop_request;
         }
      }
   );

=item client_connected => $host, $port

=item client_disconnected => $host, $port

These events are emitted whenever a client coming from C<$host:$port> connects
to your server or is disconnected from it.

=back

=head1 CACHING

Any response from the HTTP server will have C<Cache-Control> set to C<max-age=0> and
also the C<Expires> header set to the C<Date> header. Meaning: Caching is disabled.

You can of course set those headers yourself in the response, or remove them by
setting them to undef, but keep in mind that the default for those headers are
like mentioned above.

If you need more support here you can send me a mail or even better: a patch :)

=head1 AUTHOR

Robin Redeker, C<< <elmex at ta-sa.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-bs-httpd at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=AnyEvent-HTTPD>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc AnyEvent::HTTPD


You can also look for information at:

=over 4

=item * Git repository

L<http://git.ta-sa.org/AnyEvent-HTTPD.git>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=AnyEvent-HTTPD>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/AnyEvent-HTTPD>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/AnyEvent-HTTPD>

=item * Search CPAN

L<http://search.cpan.org/dist/AnyEvent-HTTPD>

=back

=head1 ACKNOWLEDGEMENTS

   Andrey Smirnov   - for keep-alive patches.
   Pedro Melo       - for valuable input in general and patches.
   Nicholas Harteau - patch for ';' pair separator support,
                      patch for allowed_methods support
   Chris Kastorff   - patch for making default headers removable
                      and more fault tolerant w.r.t. case.
   Mons Anderson    - Optimizing the regexes in L<AnyEvent::HTTPD::HTTPConnection>
                      and adding the C<backlog> option to L<AnyEvent::HTTPD>.

=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Robin Redeker, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of AnyEvent::HTTPD
